[RNMobile] Add `useScrollWhenDragging` hook

fluiddot opened the pull request:
**NOTE:** This PR points to [rnmobile/feature/drag-and-drop-block-draggable-component](https://github.com/WordPress/gutenberg/tree/rnmobile/feature/drag-and-drop-block-draggable-component), once https://github.com/WordPress/gutenberg/pull/39617 PR is merged, it should point to [rnmobile/feature/drag-and-drop](https://github.com/WordPress/gutenberg/tree/rnmobile/feature/drag-and-drop).

Closes https://github.com/wordpress-mobile/gutenberg-mobile/issues/4673.

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Add`useScrollWhenDragging` hook for allowing scroll when dragging a block.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

This hook is required for providing the drag & drop functionality for blocks, similarly as we have in the web version.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

This hook is used by the `BlockDraggable` component which notifies when the drag gesture starts and finishes. With this information and scroll events from the block list, it runs a timer that updates the scroll position via the [Reanimated's `scrollTo` function](https://docs.swmansion.com/react-native-reanimated/docs/api/nativeMethods/scrollTo). Similar to the web implementation, the scroll is controlled with a velocity value that increases by the distance between the touch position and the position where the drag started.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
1. Open a post/page.
2. Add some blocks.
3. Tap over a block for around one second.
4. Observe that the block collapses and a visual clue is shown on top of the touch position.
5. Drag the block to the top and observe that the scroll moves upwards.
6. Drag the block to the bottom and observe that the scroll moves downwards.
7. Additionally, observe that the further the block is dragged from the original position, the scroll moves faster.

## Screenshots or screencast <!-- if applicable -->
iOS|Android
--|--
<video src=https://user-images.githubusercontent.com/14905380/159924867-a410d801-29e7-4050-ba93-194e3560662e.MP4>|<video src=https://user-images.githubusercontent.com/14905380/159925231-b4a7c270-e06c-4a72-8f59-67b5cfca0b9c.mp4>

fluiddot committed:
Introduce useScrollWhenDragging hook

fluiddot committed:
Cancel animation timer on stop scrolling

fluiddot commented on packages/block-editor/src/components/block-draggable/index.native.js:
```diff
@@ -62,10 +63,19 @@ const BlockDraggableWrapper = ( { children } ) => {
 		'worklet';
 		const { contentOffset } = event;
 		scroll.offsetY.value = contentOffset.y;
+
+		draggingScrollHandler( event );
```
`useScrollWhenDragging` hook also needs to listen to scroll events, hence we connect its scroll handler with the one defined in `BlockDraggable`.

fluiddot commented on packages/block-editor/src/components/block-draggable/index.native.js:
```diff
@@ -98,9 +108,13 @@ const BlockDraggableWrapper = ( { children } ) => {
+				scrollAnimation.value = withTiming(
+					scrollOffsetTarget,
+					{ duration: SCROLL_ANIMATION_DURATION },
+					() => startScrolling( position.y )
```
In case we need to animate the scroll because the block is out of scroll view, we'll only notify the hook about starting scrolling once the animation finishes.

fluiddot commented on packages/block-editor/src/components/block-draggable/use-scroll-when-dragging.native.js:
```diff
@@ -0,0 +1,126 @@
+		offsetY.value = scroll.offsetY.value;
+		dragStartY.value = y;
+
+		animationTimer.value = 0;
+		animationTimer.value = withRepeat(
+			withTiming( 1, {
+				duration: SCROLL_INTERVAL_MS,
+				easing: Easing.linear,
+			} ),
+			-1,
+			true
+		);
```
The scroll animation needs to be running indefinitely until the drag gesture ends. For this reason, and in order of animating the scroll with Reanimated, we create a timer that repeats forever and animates `animationTimer` value from `0` to `1` every 1 second and in reverser, from `1` to `0`.

This timer is used below in `useAnimatedReaction` block for calculating the offset to be increased on every frame.

fluiddot commented on packages/block-editor/src/components/block-draggable/use-scroll-when-dragging.native.js:
```diff
@@ -0,0 +1,126 @@
+
+	const scrollOnDragOver = ( y ) => {
+		'worklet';
+		const dragDistance = Math.max(
+			Math.abs( y - dragStartY.value ) - SCROLL_INACTIVE_DISTANCE_PX,
+			0
+		);
```
The scroll is only animated if we drag the block beyond a threshold (i.e. `SCROLL_INACTIVE_DISTANCE_PX` constant).

fluiddot commented on packages/block-editor/src/components/block-draggable/use-scroll-when-dragging.native.js:
```diff
@@ -0,0 +1,126 @@
+			Math.abs( y - dragStartY.value ) - SCROLL_INACTIVE_DISTANCE_PX,
+			0
+		);
+		const distancePercentage = dragDistance / windowHeight;
```
This percentage is used for increasing the scroll velocity depending on the distance between the drag starting position and the current drag position.

fluiddot commented on packages/block-editor/src/components/block-draggable/use-scroll-when-dragging.native.js:
```diff
@@ -0,0 +1,126 @@
+				return;
+			}
+
+			const delta = Math.abs( value - previous );
+			let newOffset = offsetY.value + delta * velocityY.value;
```
The scroll is updated by calling [Reanimated's `scrollTo` function](https://docs.swmansion.com/react-native-reanimated/docs/api/nativeMethods/scrollTo), which prompts for the absolute scroll offset. In this case, we calculate the offset by adding to the current offset the pixels that have been moved on each frame (i.e. `delta * scroll velocity`).

fluiddot commented on packages/block-editor/src/components/block-draggable/use-scroll-when-dragging.native.js:
```diff
@@ -0,0 +1,126 @@
+	const startScrolling = ( y ) => {
+		'worklet';
+		stopScrolling();
+		offsetY.value = scroll.offsetY.value;
```
Since the scroll animation requires providing the absolute offset, we have to get the initial offset at the time the dragging starts. This way we can make the offset calculations using the initial offset.

fluiddot commented on packages/block-editor/src/components/block-draggable/use-scroll-when-dragging.native.js:
```diff
@@ -0,0 +1,126 @@
+
+const SCROLL_INACTIVE_DISTANCE_PX = 50;
+const SCROLL_INTERVAL_MS = 1000;
+const VELOCITY_MULTIPLIER = 5000;
```
If needed, we could tweak this value to increase the acceleration of the scroll velocity.

fluiddot committed:
Add documentation to useScrollWhenDragging hook

geriux commented on packages/block-editor/src/components/block-draggable/use-scroll-when-dragging.native.js:
```diff
@@ -0,0 +1,135 @@
+	const animatedScrollRef = useAnimatedRef();
+	animatedScrollRef( scrollRef );
+
+	const windowHeight = Dimensions.get( 'window' ).height;
```
Alternatively, we could use `useWindowDimensions` which updates automatically if the window size changes. What do you think?

fluiddot replied:
Good point, I'll replace this with `useWindowDimensions` 👍 .

fluiddot replied:
I've applied this suggestion in https://github.com/WordPress/gutenberg/pull/39705/commits/440217c70c434c784b57bf1312c95f0c43529079.

geriux reviewed (commented):
LGTM! I only had one comment, but the implementation looks good!

fluiddot committed:
Replace Dimensions with useWindowDimensions hook

fluiddot commented:
> LGTM! I only had one comment, but the implementation looks good!

Thanks for the review @geriux ❤️ ! I've just addressed the comment so the PR is ready for another review, thanks 🙇 .

geriux reviewed (approved):
LGTM! I tested this on both iOS and Android in portrait and landscape mode. 🚀 Nice work!
